transformer_engine/pytorch/csrc/extensions/cast.cpp

```diff
@@ -100,10 +100,224 @@ py::object dequantize(const py::handle &input, transformer_engine::DType otype)
+                                static_cast<DType>(out_identity_amax.dtype),
+                                out_identity_amax.shape);
+
+          NVTE_SCOPED_GIL_RELEASE({
+            nvte_quantize_v2(input_list[i].data(), out_identity.data(), quant_config_list[i],
+                             stream);
+          });
+        }
+
+        // already eligible for RHT columnwise cast fusion after the dimension check
+        if (quantizer->columnwise_usage) {
+          // Get the output columnwise data, scale_inv, and amax
+          auto out_columnwise_data = output_list[i].get_columnwise_data();
+          auto out_columnwise_scale_inv = output_list[i].get_columnwise_scale_inv();
+          // NOTE: should already be populated.
+          auto out_columnwise_amax = output_list[i].get_columnwise_amax();
+
+          // Create a wrapper for the columnwise output, as the rowwise output.
+          // The reason is due to the input `rht_output_t` is already in the transposed layout.
+          // Thus, we only need a rowwise quantization to generate the columnwise output.
+          TensorWrapper out_transpose(output_list[i].scaling_mode());
+          auto colwise_data_shape = out_columnwise_data.shape;
+          std::vector<size_t> colwise_data_shape_2d;
+          colwise_data_shape_2d.push_back(colwise_data_shape.data[0]);
+          size_t last_dim = 1;
+          for (size_t i = 1; i < colwise_data_shape.ndim; ++i) {
+            last_dim *= colwise_data_shape.data[i];
+          }
+          colwise_data_shape_2d.push_back(last_dim);
+
+          out_transpose.set_rowwise_data(out_columnwise_data.data_ptr,
+                                         static_cast<DType>(out_columnwise_data.dtype),
+                                         colwise_data_shape_2d);
+          out_transpose.set_rowwise_scale_inv(out_columnwise_scale_inv.data_ptr,
+                                              static_cast<DType>(out_columnwise_scale_inv.dtype),
+                                              out_columnwise_scale_inv.shape);
+          out_transpose.set_amax(out_columnwise_amax.data_ptr,
+                                 static_cast<DType>(out_columnwise_amax.dtype),
+                                 out_columnwise_amax.shape);
+          nvte_hadamard_transform_cast_fusion_columnwise(input_list[i].data(), out_transpose.data(),
+                                                         rht_matrix_nvte.data(),
+                                                         quant_config_list[i], stream);
+        }
+      }
+    });
+  } else {
+    NVTE_SCOPED_GIL_RELEASE({
+      for (size_t i = 0; i < num_tensors; i++) {
+        // skip this round if input is empty
+        if (input_list[i].numel() == 0) {
+          continue;
+        }
+        nvte_quantize_v2(input_list[i].data(), output_list[i].data(), quant_config_list[i], stream);
+      }
+    });
+  }
+}
+
+void multi_tensor_quantize_impl(const TensorWrapper &single_input,
```

Comment:
I think we should separate the logic for multi-tensor quantize and split + multi-tensor quantize. It's a bad sign that we're fighting against the design here: https://github.com/NVIDIA/TransformerEngine/blob/e0a47141b461a2c55f033156e94e016e7b859943/transformer_engine/pytorch/csrc/extensions/cast.cpp#L454-L460